raise more legible error if the word embedding dimensions don't match

asoderlind opened the pull request:
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
I want this PR to address this issue: https://github.com/arc53/DocsGPT/issues/550 and raise a more legible error if the embeddings have a different length of `word_embedding_dimension` than the `docsearch.index` dimension.

- **Why was this change needed?** (You can also link to an open issue here)
It's needed because not checking the different vector embedding lengths can lead to unexpected and unwanted behaviour.

- **Other information**:
The initial solution I have for this is pretty ugly so I welcome a discussion in order to find the best solution and hopefully fix the more general problem of word embedding dimension checking for all models.

dartpain commented:
Please just fix this
application/vectorstore/faiss.py:21:121: E501 Line too long (121 > 120 characters)
Good PR, important issue

dartpain commented:
Also please change tests

asoderlind commented:
Thanks for the feedback! Will fix the PR tomorrow. The linting fix is easy but what kinds of tests should I change?

dartpain commented:
Just this things please, will appreciate it a lot! Thank you!


@codecov codecov / codecov/patch

application/vectorstore/faiss.py#L16

Added line #L16 was not covered by tests

[Check warning on line 28 in application/vectorstore/faiss.py](https://github.com/arc53/DocsGPT/pull/588/files#annotation_14896088489)

@codecov codecov / codecov/patch

application/vectorstore/faiss.py#L22-L28

Added lines #L22 - L28 were not covered by tests

asoderlind commented:
Did a small refactor to make it look nicer (might be good to add more tests that also include the openAI and cohere transformers, but I don't have API keys atm)

Also added a test that raises the error in the PR. Although we make the error more readable and catch it earlier it still doesn't really fix the problem that the `index.faiss` and `huggingface_sentence-transformers/all-mpnet-base-v2` have different length embeddings. What would be the next step to actually address this when running locally? @dartpain

asoderlind commented:
Was the `index.faiss` file generated with the OpenAI transformer? @dartpain

dartpain commented:
@asoderlind yep it was generated with openai's embeddings

dartpain commented:
@asoderlind thats a good PR, just few small issues:
1: just remove unused imports
 Error: tests/test_vector_store.py:2:19: F401 `flask.Flask` imported but unused
Error: tests/test_vector_store.py:3:31: F401 `application.error.bad_request` imported but unused
Error: tests/test_vector_store.py:3:44: F401 `application.error.response_error` imported but unused
Error: Process completed with exit code 1.
2. please add sentence-transformers in our application/requirements.txt as a dependency
 FAILED tests/test_vector_store.py::test_init_local_faiss_store_huggingface - ImportError: Could not import sentence_transformers python package. Please install it with `pip install sentence-transformers`.
 
 
 Thank you so much!

asoderlind commented:
Added `sentence-transformers==2.2.2` and removed the unused imports, sorry for the oversight.

But to address the actual issue, would it be possible to generate a separate index with embeddings of length 768? Something like `index_d768.faiss` or something. It might work better when using locally with the preset ![huggingface transformer](https://huggingface.co/sentence-transformers/all-mpnet-base-v2). We can then create a mapping to the right index depending on the embeddings used.

dartpain commented:
I think I we will probably add another file, no worries, thank you for the PR
@holopin-bot @asoderlind
